### PR TITLE
Add an preliminary conversion for stablehlo.dot_general op

### PR DESCRIPTION
**Pull request by uazizTT**
This conversion only works for cases that can be directly converted for ttir.matmul op in order to support the case in MNIST examples. 

It will be enhanced further as other required ops are supported e.g. permute and broadcast_in_dim.

### CODE REVIEW

**Review by AleksKnezevic — APPROVED** [review]

File: lib/Conversion/StableHLOToTTIR/StableHLOToTTIRPatterns.cpp [hunk 1]
```diff
@@ -140,6 +140,54 @@ class StableHLOToTTIRTransposeOpConversionPattern
...
+    auto outputTensor = rewriter.create<tensor::EmptyOp>(
+        srcOp.getLoc(), outputType.getShape(), outputType.getElementType());
+
+    // This is a basic version that can only work for cases that can be directly
+    // converted to matmul. The op should be extended as other ops such as
+    // ttir.permute and ttir.broadcast_in_dim become available.
+
+    ::mlir::stablehlo::DotDimensionNumbersAttr dimensions =
+        adaptor.getDotDimensionNumbers();
+    if (dimensions.getLhsContractingDimensions().size()) {
```
Comment (mrakitaTT) [inline lib/Conversion/StableHLOToTTIR/StableHLOToTTIRPatterns.cpp:171]: You should also return failure here when this is empty. Change it to something like this (when you move this to `checkBasicLegality` function):
```
if (dimensions.getLhsContractingDimensions().size() != 1 ||
    dimensions.getLhsContractingDimensions()[0] != 1) {
  // Only regular non-transposed matmul is supported currently.
  return false;
}
```

Similar for the next `if` below.

File: lib/Conversion/StableHLOToTTIR/StableHLOToTTIRPatterns.cpp [hunk 2]
```diff
@@ -140,6 +140,54 @@ class StableHLOToTTIRTransposeOpConversionPattern
...
+    if (dimensions.getRhsContractingDimensions().size()) {
+      assert(dimensions.getRhsContractingDimensions()[0] == 0 &&
+             "ttir dot_general only supports matmul operation.");
+    }
+
+    assert(dimensions.getLhsBatchingDimensions().empty() &&
+           "ttir dot_general only supports matmul operation.");
+
+    assert(dimensions.getRhsBatchingDimensions().empty() &&
+           "ttir dot_general only supports matmul operation.");
```
[the remark this answers is not in the document]
Comment (uazizTT) [reply lib/Conversion/StableHLOToTTIR/StableHLOToTTIRPatterns.cpp:185]: I think at this stage of our compiler, if the input itself is correct, most internal errors are catastrophic i.e. the compiler cannot recover from it, that's why we use assertions. At the same time, we should provide user with a meaningful feedback about why the error occurred, hence an error message is very important. The issue with using `failure()` alone as it is now in `checkBasicLegality()` is that it doesn't inform about the cause of the failure. 

I suppose one can use `failure()` along with a debug message e.g. `LLVM_DEBUG()`, but then the user will have to open the debug log and find the cause of the error (which is fine for some cases). Hence, it is very common to use `llvm_unreachable("error message") `and `LLVM_ASSERT(condition && "error message")` are the most common LLVM based compilers. 

If the error is due to incorrect input, in that case `failure()` is a suitable option, but should include a reason so user can correct the input and try again.

I think this article best describes these points: 
http://onnx.ai/onnx-mlir/ErrorHandling.html

So to summarize I think one of the following ways is acceptable: 

1) `llvm_reachable()` or `LLVM_ASSERT()` with a brief message.
2) `op->emitError() << "message";`
3) `failure()` which is proceeded with `LLVM_DEBUG()` or `LLVM_WARN()`

File: lib/Conversion/StableHLOToTTIR/StableHLOToTTIRPatterns.cpp (hunk 1 above)
Comment (uazizTT) [reply lib/Conversion/StableHLOToTTIR/StableHLOToTTIRPatterns.cpp:171]: I will move the error checking to a separate function like `checkBasicLegality`

File: lib/Conversion/StableHLOToTTIR/StableHLOToTTIRPatterns.cpp (hunk 2 above)
[the remark this answers is not in the document]
Comment (uazizTT) [reply lib/Conversion/StableHLOToTTIR/StableHLOToTTIRPatterns.cpp]: I ended up using a combination of `op->emitError() << "message"; `and `failure()` as it produces an unrecoverable failure as well as prints a debug message to user.

[the remark this answers is not in the document]
Comment (mrakitaTT) [reply lib/Conversion/StableHLOToTTIR/StableHLOToTTIRPatterns.cpp]: Hey Usman, I totally agree that error message is also important, my main point was that assertions are not a good way of reporting errors to the user, their main usage is to catch code errors for developers. Assertions -> things you expect to never happen, otherwise it’s a bug. Errors -> things you expect to happen and come from input to the program, which is the case here. Even the link you posted supports this, you were asserting during input graph validation phase.

I like your investigation about proper error handling in llvm and the solution that you ended up using, we already use `op->emitError()` in other passes, it is basically a wrapper around `return failure()` which also prints error message.

File: lib/Conversion/StableHLOToTTIR/StableHLOToTTIRPatterns.cpp [hunk 3]
```diff
@@ -133,19 +133,103 @@ class StableHLOToTTIRTransposeOpConversionPattern
...
+    return success();
+  }
+
+  bool
+  checkBasicLegality(mlir::stablehlo::TransposeOp &srcOp,
+                     mlir::stablehlo::TransposeOp::Adaptor &adaptor) const {
+
+    if (adaptor.getPermutation().size() != 2) {
+      srcOp->emitError()
+          << "TTIR only supports only two dimensional transposeOp.";
```
Comment (mrakitaTT) [inline lib/Conversion/StableHLOToTTIR/StableHLOToTTIRPatterns.cpp:158]: Delete one `only` extra

File: lib/Conversion/StableHLOToTTIR/StableHLOToTTIRPatterns.cpp (hunk 1 above)
Comment (mrakitaTT) [reply lib/Conversion/StableHLOToTTIR/StableHLOToTTIRPatterns.cpp]: This wasn't fixed yet when you moved the code to `checkBasicLegality`

File: lib/Conversion/StableHLOToTTIR/StableHLOToTTIRPatterns.cpp [hunk 4]
```diff
@@ -133,19 +133,103 @@ class StableHLOToTTIRTransposeOpConversionPattern
...
+  checkBasicLegality(mlir::stablehlo::DotGeneralOp &srcOp,
+                     mlir::stablehlo::DotGeneralOp::Adaptor &adaptor) const {
+
+    ::mlir::stablehlo::DotDimensionNumbersAttr dimensions =
+        adaptor.getDotDimensionNumbers();
+
+    if (dimensions.getLhsContractingDimensions().size()) {
+      if (dimensions.getLhsContractingDimensions()[0] != 1) {
+        srcOp->emitError()
+            << "ttir dot_general only supports matmul operation.";
```
Comment (mrakitaTT) [inline lib/Conversion/StableHLOToTTIR/StableHLOToTTIRPatterns.cpp:208]: TTIR doesn't have a `dot_general` function, I think you should change this message to something like:
```
Only regular non-transposed matmul is supported currently in TTIR.
```

File: lib/Conversion/StableHLOToTTIR/StableHLOToTTIRPatterns.cpp [hunk 5]
```diff
@@ -133,19 +133,110 @@ class StableHLOToTTIRTransposeOpConversionPattern
     tensor::EmptyOp outputTensor = rewriter.create<tensor::EmptyOp>(
         srcOp.getLoc(), outputType.getShape(), outputType.getElementType());
 
-    assert(adaptor.getPermutation().size() == 2 &&
-           "TTIR only supports only two dimensional transposeOp.");
+    if (!checkBasicLegality(srcOp, adaptor)) {
+      LLVM_DEBUG(llvm::dbgs() << "Failed legality checks in the handling of Op "
+                              << srcOp->getName());
```
Comment (mrakitaTT) [inline lib/Conversion/StableHLOToTTIR/StableHLOToTTIRPatterns.cpp:138]: Maybe I'm missing something but what's the point of this? We will already throw the error with the specific message what failed.

File: lib/Conversion/StableHLOToTTIR/StableHLOToTTIRPatterns.cpp [hunk 6]
```diff
@@ -133,19 +133,110 @@ class StableHLOToTTIRTransposeOpConversionPattern
...
+      return failure();
+    }
 
     rewriter.replaceOpWithNewOp<mlir::tt::ttir::TransposeOp>(
         srcOp, outputTensor.getType(), Value(adaptor.getOperand()),
         Value(outputTensor), adaptor.getPermutation()[0],
         adaptor.getPermutation()[1],
+
```
Comment (mrakitaTT) [inline lib/Conversion/StableHLOToTTIR/StableHLOToTTIRPatterns.cpp:146]: Just curious, is the clang formatter inserting this empty line for you or..?

File: lib/Conversion/StableHLOToTTIR/StableHLOToTTIRPatterns.cpp (hunk 1 above)
Comment (mrakitaTT) [reply lib/Conversion/StableHLOToTTIR/StableHLOToTTIRPatterns.cpp]: In any case it is a good practice to check for the size of the external container before indexing it

File: lib/Conversion/StableHLOToTTIR/StableHLOToTTIRPatterns.cpp (hunk 5 above)
Comment (uazizTT) [reply lib/Conversion/StableHLOToTTIR/StableHLOToTTIRPatterns.cpp:138]: The purpose of this is to start documenting the flow of the program execution in a log file. I just setup DEBUG_TYPE so we can all start logging that aids in debugging.

File: lib/Conversion/StableHLOToTTIR/StableHLOToTTIRPatterns.cpp (hunk 6 above)
Comment (uazizTT) [reply lib/Conversion/StableHLOToTTIR/StableHLOToTTIRPatterns.cpp:146]: Yes something weird is happening, manually deleting now.

File: lib/Conversion/StableHLOToTTIR/StableHLOToTTIRPatterns.cpp (hunk 5 above)
Comment (mrakitaTT) [reply lib/Conversion/StableHLOToTTIR/StableHLOToTTIRPatterns.cpp:138]: Isn't it better then to catch the errors in one place somewhere top level and to just log their message there? This seems kind of redundant.

File: lib/Conversion/StableHLOToTTIR/StableHLOToTTIRPatterns.cpp [hunk 7]
```diff
@@ -186,6 +234,13 @@ void addTransposeOpsConversionPatterns(MLIRContext *ctx,
   patterns.add<StableHLOToTTIRTransposeOpConversionPattern>(typeConverter, ctx);
 }
 
+void dotGeneralOpsConversionPatterns(MLIRContext *ctx,
```
[the remark this answers is not in the document]
Comment (mrakitaTT) [reply lib/Conversion/StableHLOToTTIR/StableHLOToTTIRPatterns.cpp]: I guess you didn't like my suggestion to rename to `addMatmulOpsConversionPatterns` 😅 I won't push you but I just wanted to elaborate more about my thinking here. My thinking was that we should group the patterns insertion into some logical groups depending on the resulting class of ops. For example element-wise ops, data-movement ops, matmul ops, etc. If the function is named after specific StableHLO function like `addDotGeneralOpsConversionPatterns` then it can contain only addition of that one pattern, so that almost defeats the purpose of even having that function.

File: lib/Conversion/StableHLOToTTIR/StableHLOToTTIRPatterns.cpp (hunk 5 above)
Comment (uazizTT) [reply lib/Conversion/StableHLOToTTIR/StableHLOToTTIRPatterns.cpp:138]: Ok I should have explained it in detail, my purpose of LLVM_DEBUG() at this stage was NOT to log errors but instead to inform about the flow of the program execution to aid in debugging e.g. we should have LLVM_DEBUG() at each entry and exit points of functions so developer can tell the flow of execution just by looking at the logs rather than having to gdb through the code every time. 

But I should have introduced this logging in separate PR so not to bundle it up with error checking, so I am removing it and will open a separate PR for it.

@nsmithtt Sure I will replace it.

File: lib/Conversion/StableHLOToTTIR/StableHLOToTTIRPatterns.cpp (hunk 7 above)
[the remark this answers is not in the document]
Comment (uazizTT) [reply lib/Conversion/StableHLOToTTIR/StableHLOToTTIRPatterns.cpp]: Thanks for elaborating your point of view, I don't a strong preference for the patterns function name, just that I think it should be closer to `stablehlo` names so it is easier to identify the op itself from the pattern function name. 

In this case `stablehlo` doesn't have `matmul` op, but rather it has the `stablehlo.dot` and `stablehlo.dot_general` op, from which `dot_general` op is the broader op, so thought to keep it to `addDotOpsConversionPatterns`. But if you have a strong preference to naming it `addMatmulOpsConversionPatterns` then I can change it.

[the remark this answers is not in the document]
Comment (nsmithtt) [reply lib/Conversion/StableHLOToTTIR/StableHLOToTTIRPatterns.cpp]: I think it fundamentally boils down to, are we naming these functions based on the operator names that we're lowering _from_ or the names that we're lowering _to_?  I'm somewhat indifferent either way, but whatever we pick we should be consistent.

[the remark this answers is not in the document]
Comment (mrakitaTT) [reply lib/Conversion/StableHLOToTTIR/StableHLOToTTIRPatterns.cpp]: > I think it fundamentally boils down to, are we naming these functions based on the operator names that we're lowering from or the names that we're lowering to?

I wanted to make it more universal, to group the ops that lower in the same way (using same OpConversionPattern) and which belong to the same general "class" of ops (eltwise, binary, reduce, matmul, data movement, etc). One goal was obviously to split the code into organized functions, and another was so that if someone is working on ONNX->TTIR conversion and wants to see how for example matmul op is converted for StableHLO, they can easily locate it by finding addMatmulOps function instead of knowing how exactly it is called in StableHLO. I think that matmul is much more recognized name in ML world than dot_general, which is why I wanted to avoid using specific names from any dialect but to stick to some more universal names. Also if we use specific names from some dialect then we condemn these functions to contain only one or two ops, which would explode the number of functions. In any case, I wouldn't name these functions based on the operator names that we're lowering _from_ because sometimes we are lowering couple of ops into one target op, and there it is natural to call the function based on targeting op.

Now I made the whole philosophy about totally unimportant topic, I apologize for that, I have OCD about these little things 😅

File: lib/Conversion/StableHLOToTTIR/StableHLOToTTIRPatterns.cpp (hunk 5 above)
Comment (mrakitaTT) [reply lib/Conversion/StableHLOToTTIR/StableHLOToTTIRPatterns.cpp:138]: That sounds okay, though you should probably check with rest of uhuru team about that (if you already haven't) to make sure the whole project is using the same logging infrastructure and to agree on what tools/infra we want to use for that.

File: lib/Conversion/StableHLOToTTIR/StableHLOToTTIRPatterns.cpp (hunk 7 above)
[the remark this answers is not in the document]
Comment (nsmithtt) [reply lib/Conversion/StableHLOToTTIR/StableHLOToTTIRPatterns.cpp]: I agree, I think Usman changed it to read matmul so let's move forward with that.

File: lib/Conversion/StableHLOToTTIR/StableHLOToTTIRPatterns.cpp (hunk 5 above)
Comment (nsmithtt) [reply lib/Conversion/StableHLOToTTIR/StableHLOToTTIRPatterns.cpp:138]: @uazizTT, rest of review looks good to me, but perhaps we can use `rewriter.notifyMatchFailure` here before merging?

Comment (uazizTT) [reply lib/Conversion/StableHLOToTTIR/StableHLOToTTIRPatterns.cpp]: Updated to use `rewriter.notifyMatchFailure`

Comment (mrakitaTT) [reply lib/Conversion/StableHLOToTTIR/StableHLOToTTIRPatterns.cpp]: Just to make sure we are all in clear for the future changes, @nsmithtt do you think it is okay that we are using `srcOp->emitError()` inside `checkBasicLegality` function and then also using `rewriter.notifyMatchFailure` in case `checkBasicLegality` returns false, or do you think it would be better if we change this to something like:
```
LogicalResult legalityResult = checkBasicLegality(...);
if (!legalityResult.succeeded()) {
  return legalityResult;
}

...

LogicalResult checkBasicLegality(...) const {
  if (...) {
    return rewriter.notifyMatchFailure(srcOp, "error message");
  }

  return success();
}
```

Latter seems a bit cleaner to me.

Comment (uazizTT) [reply lib/Conversion/StableHLOToTTIR/StableHLOToTTIRPatterns.cpp]: Yeah I think we should use one API, let me update it for all conversions.

Comment (nsmithtt) [reply lib/Conversion/StableHLOToTTIR/StableHLOToTTIRPatterns.cpp]: I agree, latter is cleaner to me too. @uazizTT, sorry for bikeshedding your review, but since this is fresh code I think it's good to establish the way we want to organize this kind of stuff.  In the future this code will definitely be copy-pasted so it's good to establish best practices at the start.

Comment (mrakitaTT) [reply lib/Conversion/StableHLOToTTIR/StableHLOToTTIRPatterns.cpp]: > @uazizTT, sorry for bikeshedding your review, but since this is fresh code I think it's good to establish the way we want to organize this kind of stuff. In the future this code will definitely be copy-pasted so it's good to establish best practices at the start.

+1 I wanted to say the same thing :)

Comment (uazizTT) [reply lib/Conversion/StableHLOToTTIR/StableHLOToTTIRPatterns.cpp]: OK I refactored and push the changes.

**Review by mrakitaTT — APPROVED** [review]

**Review by nsmithtt — APPROVED** [review]
Looks good, thanks @uazizTT!
